### PR TITLE
Improve Products block Attributes Filter Inspector Controls

### LINKED ISSUE #8135: Product Query: Improve UX of the attributes filter inspector controls
The attribute filters inspector controls currently use a regular `FormTokenField`. The latest design iteration by @vivialice has a much better UX flow. We should implement the flow according to this spec:

Figma: UmuaN1XXqCa6myOHkb0ZWy-fi-1313%3A119836&t=a12roEsuCt8ErZjq-1

### PR DESCRIPTION
**Pull request by sunyatasattva**
This PR is meant to improve the UI and UX behind the Attributes filter within the Inspector Controls of the “Products (Beta)“ block. While doing so, I wanted to use the original Attribute selector component we had (`AttributeTermControl`) instead of creating a new, almost same, one. As such, this PR does a bunch of things to improve and extend that component as well.

In particular:

* It refactors the `useProductAttributes` hook that was created for the draft filter. The refactored hook is now available in the shared code, instead of just the `product-query` block directory, and has been modified to achieve the following:
   *  It fetches both terms AND attributes via the API (previously, we got the attributes from the settings, but we'd get
partial objects compared to the API? Maybe a follow-up to this could be to check why the attributes stored in the settings are incomplete?).
  * Make sure the return values match the ones expected from search items.
* Improves the functionality of the `SearchListControl` component with the following additions:
  * Allow the search input to also be a Token-based input.
  * Allow the search control to search within collapsed properties (previously, the search control would only return results if properties were uncollapsed).
  * Use core `CheckboxControl` instead of radio buttons for items having children (besides using the core component, it also helps because `CheckboxControl` includes the `indeterminate` state, which was required by the design spec).
    * This checkbox now acts more intuitively: If indeterminate or unchecked, it will check all children, otherwise, it will uncheck all the children.
  * Refactor styles for the component and also adds some more semantic classes.
  * Enable correct keyboard navigation through the component.
  * Preload attributes and terms.
  * Improves type signatures related to this component.
* It adds this component to the Inspector Controls, plus a link to manage attributes in the store.
* Implement type signatures for the `ProductAttributeTermControl` component and remove the `withAttributes` HOC wrapping it, in favor of the `useProductAttributes` hook.

Closes woocommerce/woocommerce-blocks#8135

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

Screenshot 2023-03-01 at 00 31 19

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

##### Simple happy path

1. Add a Products (Beta) block to your page.
2. Open the Inspector Controls and add the “Product Attributes” advanced filter.
3. The Attributes selector component should show up in an uninteractable loading state.
4. Once loading finishes, the Attributes should appear, collapsed.
5. Uncollapse a section
6. Select a few attributes.
7. Publish your page.
8. Check on the frontend to see that the correct products are displayed.

##### Checkbox behavior

1. Repeat steps 1–5 above.
2. Click on some terms checkboxes.
3. Notice that the parent checkbox will be in the indeterminate state (`[-]`).
4. Select all the children.
5. Notice parent checkbox will be checked.
6. Uncheck the parent checkbox.
7. All the children should be unchecked.
8. Check the parent checkbox.
9. All the children should be checked.
10. Uncheck a few children.
11. Check the parent checkbox.
12. All children should be checked.

##### Token input behavior

1. Repeat steps 1–4 above.
2. Start typing something within the input.
3. The checklist should be correctly filtered showing breadcrumbs.
4. Try the behavior for both collapsed and uncollapsed states.
5. Every time a checkbox is checked, it should have the corresponding token in the input field and viceversa.
6. Tokens can be correctly removed from the input field and this will uncheck the corresponding checkbox.

##### Notes

Note that the `SearchListControl` is used in other components which may or may not be covered by E2E tests, and this PR refactors it. I tried my best to avoid any possible regression, but please do double check with components you know. Some examples are Filter By Attribute block or Featured Product/Category.

##### Known issues

1. https://github.com/woocommerce/woocommerce/issues/42497
2. https://github.com/woocommerce/woocommerce-blocks/issues/8530
3. There is a TypeScript error in `item.tsx#L118` due to wrong Gutenberg types. Will update them upstream.
4. When searching for a full text of an item and pressing <kbd>Enter</kbd> the token will be added but the text will remain. This is due to an issue upstream.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Products block: improved UX and UI for the Product Attributes advanced filter.

### CODE REVIEW

**Comment by sunyatasattva** [comment]
@nefeline This should now be ready for review 🙏

**Comment by nefeline** [comment]
> @nefeline This should now be ready for review 🙏

Awesome! I'm getting started now 🙌

**Review by nefeline — CHANGES_REQUESTED** [review]
Great work! It's fantastic to see all of these improvements in place 🎉 

I have encountered some errors, all of them detailed below:

## Test results:

### Simple happy path

> Add a Products (Beta) block to your page.
> Open the Inspector Controls and add the “Product Attributes” advanced filter.
> The Attributes selector component should show up in an uninteractable loading state.

✅ Confirmed the attributes selector component shows up as expected.

> Once loading finishes, the Attributes should appear, collapsed.

✅  Confirmed the attributes show up collapsed:

Screenshot 2023-03-03 at 17 58 03

> Uncollapse a section
> Select a few attributes.

✅  Confirmed it's possible to uncollapse the sections and select the terms:

Screenshot 2023-03-03 at 18 02 07


> Publish your page.
> Check on the frontend to see that the correct products are displayed.

❌  At this stage, on the FE, a PHP notice is triggered, and the page is blank:

`Notice: Undefined index: taxonomy in /Users/patriciahillebrandt/Woo/plugins/woocommerce-blocks/src/BlockTypes/ProductQuery.php on line 301`

I was able to circumvent this error by doing a quick refactor to the [`get_product_attributes_query`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/ProductQuery.php#L297-L322) method as the `$attributes` variable received has a different format than the one expected (e.g. it doesn't include the 'taxonomy' name for example.):

Screenshot 2023-03-03 at 17 43 49

This is the quick fix I put in place to circumvent it:

```php
private function get_product_attributes_query( $attributes = array() ) {
		$tax_queries = array();
		foreach ( $attributes as $attribute ) {
			$term_id = $attribute['termId'];
			$term = get_term( $term_id );

			if ( $term && ! is_wp_error( $term ) ) {
				$taxonomy = $term->taxonomy;
				if ( ! array_key_exists( $taxonomy, $tax_queries ) ) {
					$tax_queries[ $taxonomy ] = array(
						'taxonomy' => $taxonomy,
						'field'    => 'term_id',
						'terms'    => array( $term_id ),
						'operator' => 'IN',
					);
				} else {
					$tax_queries[ $taxonomy ]['terms'][] = $term_id;
				}
			}
		}

		return array(
			'tax_query' => array_values( $tax_queries ),
		);
}
```

This is not necessarily the final solution; another alternative could be to ensure the information received via the `$attributes` variable matches the expected by this method instead.

With this error out of the way, I was able to confirm the FE correctly displays all relevant products (in this test, all products with the colors blue and gray are listed, even though, in some cases, the images of the products displayed are matching the product variation options available and not necessarily the filtered colors (and this is pre-existing).

🗒️ Sidenote: do we have an issue opened for this already? It feels wrong to display a red hoodie if the filter is deliberately specifying only blue product variations should be displayed (I noticed this same behavior with the attribute filter before).

Screenshot 2023-03-03 at 18 14 35

### Checkbox behavior

> Repeat steps 1–5 above.
> Click on some terms checkboxes.
> Notice that the parent checkbox will be in the indeterminate state ([-]).

✅  Confirmed the parent checkbox is in the indeterminate state:

Screenshot 2023-03-03 at 19 00 24

> Select all the children.
> Notice parent checkbox will be checked.

✅  Confirmed the parent checkbox is checked when all children are selected:

Screenshot 2023-03-03 at 19 01 45

> Uncheck the parent checkbox.
> All the children should be unchecked.

✅  Confirmed all children are unchecked, but with one limitation:

❌ When a combination of terms from different taxonomies is selected, if you uncheck all of them for one category in particular, the terms are removed for all categories instead (not just the selected one):

https://user-images.githubusercontent.com/15730971/222792297-3c6e572e-09c8-4e69-aeab-b8c668290103.mov

> Check the parent checkbox.
> All the children should be checked.

✅  Confirmed all children are checked when the parent checkbox is clicked.

> Uncheck a few children.
> Check the parent checkbox.
> All children should be checked.

✅  Confirmed this works as expected.

### Token input behavior
> Repeat steps 1–4 above.
> Start typing something within the input.
> The checklist should be correctly filtered showing breadcrumbs.

✅  Confirmed the checklist is correctly filtered and its showing the breadcrumbs:

Screenshot 2023-03-03 at 18 31 58

❌ When the name of a term is typed, the token is added as expected, but the typed text is still displayed on the side (you have to manually remove the text to be able to search for the next term):

https://user-images.githubusercontent.com/15730971/222791473-d2a6774b-bd24-4c0b-865b-ba11251e4533.mov

> Try the behavior for both collapsed and uncollapsed states.

✅  Confirmed it works as expected with both collapsed and uncollapsed states.

> Every time a checkbox is checked, it should have the corresponding token in the input field and vice-versa.

✅ Confirmed it works as expected.

> Tokens can be correctly removed from the input field and this will uncheck the corresponding checkbox.

✅  Confirmed it works as expected.

**Comment by sunyatasattva** [comment]
Hey @nefeline!

Thanks a lot for the extremely helpful review!

> ❌ At this stage, on the FE, a PHP notice is triggered, and the page is blank:

Ouch, silly me for not noticing this. I did some last minute refactoring of the types and I didn't check this thing. Thank you for providing also an alternative solution, but as you mentioned, I fixed it upstream so that the function receives the data it expects.

> 🗒️ Sidenote: do we have an issue opened for this already? It feels wrong to display a red hoodie if the filter is deliberately specifying only blue product variations should be displayed (I noticed this same behavior with the attribute filter before).

Yea, this was pre-existing in all our implementations (including the front-end filter). We do not have an issue for this AFAIK. Also I'm not sure how the products work under the hood. If you go to the product page, you'll see that “red hoodie” is the default picture. Pictures are then assigned to product “variations”, not attributes themselves. So I guess this would need a deep dive into how all these things work to make it happen. Definitely weird, though, I would agree.

> ❌ When a combination of terms from different taxonomies is selected, if you uncheck all of them for one category in particular, the terms are removed for all categories instead (not just the selected one):

Fixed. My math skills faltered 😅 

> ❌ When the name of a term is typed, the token is added as expected, but the typed text is still displayed on the side (you have to manually remove the text to be able to search for the next term):

Unfortunately this is a known issue I mentioned in the OP:

> When searching for a full text of an item and pressing Enter the token will be added but the text will remain. This is due to an issue upstream.

There was a workaround I had found, but it had its own weird side-effects unfortunately (the text would be removed also when clicking on checkboxes in my workaround). Having to decide between the lesser evil, I chose this route, which, in the end, preserves the text typed by the user which can be useful when you are doing a partial search and not a full match. It's not ideal, but it was either this, my other workaround, or rebuilding the Token input from scratch.

I'll try and contribute upstream for this.

File: assets/js/base/context/hooks/use-product-attributes.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import { formatError } from '@woocommerce/base-utils';
+
+export default function useProductAttributes( shouldLoadAttributes: boolean ) {
+	const [ errorLoadingAttributes, setErrorLoadingAttributes ] =
+		useState< Awaited< ReturnType< typeof formatError > > | null >( null );
+	const [ isLoadingAttributes, setIsLoadingAttributes ] = useState( false );
+	const [ productsAttributes, setProductsAttributes ] = useState<
+		AttributeWithTerms[]
+	>( [] );
+	const hasLoadedAttributes = useRef( false );
```
Comment (nefeline) [inline assets/js/base/context/hooks/use-product-attributes.ts:20]: Nit: any particular reason for relying on `useRef` instead of the state hook here?

**Review by nefeline — APPROVED** [review]
Thanks for addressing the feedback! Regarding the code, I've left just one minor question & optional change, other than that this is looking sharp & I'm pre-approving!

> Thanks a lot for the extremely helpful review!

Anytime! 🙌 

> I fixed it upstream so that the function receives the data it expects.

✅  Awesome, thanks! I can confirm the notice is now gone, and the correct products are displayed on the FE with this fix in place.

> Fixed. My math skills faltered 😅

✅  😄 It happens to the best of us, thanks for addressing the issue! A can confirm when a combination of terms from different taxonomies is selected; if you uncheck all of them for one category in particular, the correct/expected terms are now removed, and not all of them.

> There was a workaround I had found, but it had its own weird side-effects unfortunately (the text would be removed also when clicking on checkboxes in my workaround). Having to decide between the lesser evil, I chose this route, which, in the end, preserves the text typed by the user which can be useful when you are doing a partial search and not a full match. It's not ideal, but it was either this, my other workaround, or rebuilding the Token input from scratch. I'll try and contribute upstream for this.

Gotcha: sorry I missed the description of this limitation in particular. The path chosen here makes sense, thanks for clarifying!

> Yea, this was pre-existing in all our implementations (including the front-end filter). We do not have an issue for this AFAIK. Also I'm not sure how the products work under the hood. If you go to the product page, you'll see that “red hoodie” is the default picture. Pictures are then assigned to product “variations”, not attributes themselves. So I guess this would need a deep dive into how all these things work to make it happen. Definitely weird, though, I would agree.

Thanks 👍 ! I'm opening a separate issue so we can keep track & investigate later on when time permits.

File: assets/js/base/context/hooks/use-product-attributes.ts (hunk 1 above)
Comment (sunyatasattva) [reply assets/js/base/context/hooks/use-product-attributes.ts:20]: @nefeline Oh, yes. The `useState` hook cannot be used here: this variable is used to track whether attributes are loaded successfully and so to load them once and not at every re-render. Using `useState` would get into an infinite loop as the state is reset to the initial value at each render.

Comment (nefeline) [reply assets/js/base/context/hooks/use-product-attributes.ts:20]: Gotcha, thanks for clarifying!

**Pull request merged**